_posts/gRPC-Go performance improvements.md

```diff
@@ -0,0 +1,148 @@
+
+While running experiments on VMs in different continents we realized that every once in awhile a rogue unnaturally fast ping-ack at the right time (really the wrong time) would cause our window sizes to go up. This happens because such a ping-ack would cause us to notice a decreased RTT and calculate a high bandwidth value. Now if that sample of bytes was greater than 2/3rd of our window then we would increase the window sizes. However, this ping ack was an aberration and shouldn't have changed our perception of the network RTT al together. Therefore, we keep a running average of the RTTs we note weighted by a constant rather than the total number of samples to heed more to recent RTTs and less to the ones in past. It is important because networks might change over time.
+
+During our implementation we experimented with different parameters like instead of putting our window size to twice the bdp sample we put it to 4/3 times the bdp sample. Note although this is more accurate (since the noted bdp was smaller than or equal to 1.5 the real bandwidth and ideally we want to be twice the real bandwidth so multiplying the sample with 4/3 should put us there), it's not the best option to choose since the growth is slow and it takes some tens of RPCs to ramp up.
+
+Given that we're always bound by the flow control of TCP which for most cases is upper bounded at 4MB, we bound the growth of our window sizes by the same number: 4MB.
+
+BDP estimation and dynamically adjusting window sizes is turned-on by default and can be turned off by setting values manually for connection and/or stream window sizes.
+
+
+## Near-future efforts
+
+
+### Flush syscalls
+
+We noticed a bug in our transport layer which causes us to make a flush syscall for every data frame we write, even if the same goroutine has more data to send. We can batch a lot of these writes to use only one flush. This in fact will not be a big change to the code itself. The performance gain noted by this change is: // numbers
+
+In our efforts to get rid of unnecessary flushes we recently combined the headers and data write for unary and server streaming RPCs to one flush on the client-side. // PR
+
+Another related idea proposed by one of our users (yay community!) was to combine a server response to a unary RPC into one flush. We are currently looking into that as well.
```

Comment:
A related idea proposed by (link name here), a community contributor, was ..

Since we named ejona above, we should name the user who proposed this as well. "yay community!"  does not come across as sincere appreciation.